dora/core/client/fs/src/main/java/alluxio/client/file/dora/WorkerLocationPolicy.java

```diff
@@ -67,32 +71,64 @@ public List<BlockWorkerInfo> getPreferredWorkers(List<BlockWorkerInfo> blockWork
   private static class ConsistentHashProvider {
     private static final HashFunction HASH_FUNCTION = murmur3_32_fixed();
     private static final int MAX_ATTEMPTS = 100;
-    private List<BlockWorkerInfo> mLastWorkerInfos = ImmutableList.of();
-    private NavigableMap<Integer, BlockWorkerInfo> mActiveNodesByConsistentHashing;
-
-    private volatile long mLastUpdatedTimestamp = 0L;
-
-    private final AtomicBoolean mNeedUpdate = new AtomicBoolean(false);
-
-    private static final long WORKER_INFO_UPDATE_INTERVAL_MS = 1000L;
-
+    private static final long WORKER_INFO_UPDATE_INTERVAL_NS = Constants.SECOND_NANO;
+    private volatile List<BlockWorkerInfo> mLastWorkerInfos = ImmutableList.of();
+    // Can only be null before the first call to refresh
+    @Nullable
+    private volatile NavigableMap<Integer, BlockWorkerInfo> mActiveNodesByConsistentHashing;
+    private final Semaphore mInitLock = new Semaphore(1);
+
+    // Must use System.nanoTime to ensure monotonic increment
+    private final AtomicLong mLastUpdatedTimestamp = new AtomicLong(System.nanoTime());
+
+    /**
+     * Initializes or refreshes the worker list using the given list of workers and number of
+     * virtual nodes.
+     */
     public void refresh(List<BlockWorkerInfo> workerInfos, int numVirtualNodes) {
-      // check if we need to update worker info
-      if (mLastUpdatedTimestamp <= 0L
-          || System.currentTimeMillis() - mLastUpdatedTimestamp > WORKER_INFO_UPDATE_INTERVAL_MS) {
-        mNeedUpdate.set(true);
+      Preconditions.checkArgument(!workerInfos.isEmpty(),
```

Comment:
i feel instead of throwing `IllegalArgumentEx` here, a checked exception is more explicit for callers (easier to catch and handle).